deps/status/src/checkLockfilesUpToDate.ts

```diff
@@ -0,0 +1,380 @@
+  satisfiesPackageManifest,
+} from '@pnpm/lockfile.verification'
+import { globalWarn, logger } from '@pnpm/logger'
+import { arrayOfWorkspacePackagesToMap } from '@pnpm/get-context'
+import { parseOverrides } from '@pnpm/parse-overrides'
+import { type WorkspacePackages } from '@pnpm/resolver-base'
+import {
+  type DependencyManifest,
+  type Project,
+  type ProjectId,
+  type ProjectManifest,
+} from '@pnpm/types'
+import { findWorkspacePackages } from '@pnpm/workspace.find-packages'
+import { loadPackagesList, updatePackagesList } from '@pnpm/workspace.packages-list-cache'
+import { readWorkspaceManifest } from '@pnpm/workspace.read-manifest'
+import { assertLockfilesEqual } from './assertLockfilesEqual'
+import { statManifestFile } from './statManifestFile'
+
+export type CheckLockfilesUpToDateOptions = Pick<Config,
+| 'allProjects'
+| 'autoInstallPeers'
+| 'catalogs'
+| 'excludeLinksFromLockfile'
+| 'linkWorkspacePackages'
+| 'hooks'
+| 'peersSuffixMaxLength'
+| 'rootProjectManifest'
+| 'rootProjectManifestDir'
+| 'sharedWorkspaceLockfile'
+| 'workspaceDir'
+>
+
+export async function checkLockfilesUpToDate (opts: CheckLockfilesUpToDateOptions): Promise<void> {
+  const {
+    allProjects,
+    autoInstallPeers,
+    catalogs,
+    excludeLinksFromLockfile,
+    linkWorkspacePackages,
+    rootProjectManifest,
+    rootProjectManifestDir,
+    sharedWorkspaceLockfile,
+    workspaceDir,
+  } = opts
+
+  const rootManifestOptions = rootProjectManifest && rootProjectManifestDir
+    ? getOptionsFromRootManifest(rootProjectManifestDir, rootProjectManifest)
+    : undefined
+
+  if (allProjects && workspaceDir) {
+    const packagesList = await loadPackagesList(workspaceDir)
+    if (!packagesList) {
+      throw new PnpmError('RUN_CHECK_DEPS_NO_CACHE', 'Cannot check whether dependencies are outdated', {
+        hint: 'Run `pnpm install` to create the cache',
+      })
+    }
+
+    if (!equals(
+      filter(value => value != null, packagesList.catalogs ?? {}),
+      filter(value => value != null, catalogs ?? {})
+    )) {
+      throw new PnpmError('RUN_CHECK_DEPS_OUTDATED', 'Catalogs cache outdated', {
+        hint: 'Run `pnpm install` to update the catalogs cache',
+      })
+    }
+
+    const currentProjectRootDirs = allProjects.map(project => project.rootDir).sort()
+    if (!equals(packagesList.projectRootDirs, currentProjectRootDirs)) {
+      throw new PnpmError('RUN_CHECK_DEPS_WORKSPACE_STRUCTURE_CHANGED', 'The workspace structure has changed since last install', {
+        hint: 'Run `pnpm install` to update the workspace structure and dependencies tree',
+      })
+    }
+
+    const allManifestStats = await Promise.all(allProjects.map(async project => {
+      const manifestStats = await statManifestFile(project.rootDir)
+      if (!manifestStats) {
+        // this error should not happen
+        throw new Error(`Cannot find one of ${MANIFEST_BASE_NAMES.join(', ')} in ${project.rootDir}`)
```

Comment:
> It would just mean that an installation is needed, no?

It means that pnpm classified the directory as a project dir without a recognized manifest (one of `package.{json,json5,yaml}`). It is a programmer error. The current version of the codebase cannot produce this invariant, but future edits may accidentally introduce it.